RFC: use upstream ARM Trusted Firmware for rk3588

stintel opened the pull request:
Build-tested on x86_64
Runtime-tested on rockchip/armv8 on Radxa ROCK 5B+

Please check individual commit messages.

stintel committed:
arm-trusted-firmware-rockchip: bump to 2.12.1 LTS

And remove no-warn-rwx-segments hack, as this is no longer needed and
actually causes build to fail. See also commit c117d7a37a74
("include/trusted-firmware-a.mk: remove the no-warn-rwx-segments hack").

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel committed:
arm-trusted-firmware-rockchip: support rk3588

ATF supports rk3588 since version 2.12, so let's enable it in
arm-trusted-firmware-rockchip.

We still need the TPL to initialize RAM, but keeping rk3588 in rkbin
would cause a duplicate package name, so rename rk3588 to rk3588-tpl in
rkbin.

Finally, point uboot-rockchip for rk3588 to bl31 built by this package,
and add a dependency on the new trusted-firmware-a-rk3588-tpl pacakge.

While this doesn't necessarily add features, we now no longer rely on
the rk3588_bl31_v1.45.elf blob provided by Rockchip, which is always a
good thing.

Tested on Radxa ROCK 5B+.

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel commented:
@1715173329 @BWhitten @RadxaNaoki @blocktrron

stintel commented:
@Manawyrm

1715173329 commented on package/boot/arm-trusted-firmware-rockchip/Makefile:
```diff
@@ -67,6 +70,22 @@ ifeq ($(BUILD_VARIANT),rk3399)
 endif
 
+ifeq ($(BUILD_VARIANT),rk3588)
+  TPL_FILE:=rk3588_ddr_lp4_2112MHz_lp5_2400MHz_v1.18.bin
```
Maybe rename rk3588 to rk3588-tpl in rkbin so we don't need to download this separately?

stintel replied:
Great idea!

1715173329 commented on package/boot/rkbin/Makefile:
```diff
@@ -57,19 +57,12 @@ define Trusted-Firmware-A/rk3568-e25
-	rk3588
+	rk3568-e25
 
 TFA_MAKE_FLAGS+= \
```
I guess this should be removed too, this package even does not require compile

stintel replied:
Not really related to this PR, but I guess it can go in a separate commit.

RadxaNaoki commented on package/boot/arm-trusted-firmware-rockchip/Makefile:
```diff
@@ -7,10 +7,10 @@
 include $(TOPDIR)/rules.mk
 
-PKG_VERSION:=2.10
+PKG_VERSION:=2.12
```
There is 2.12.1
https://git.trustedfirmware.org/plugins/gitiles/TF-A/trusted-firmware-a.git/+/refs/tags/lts-v2.12.1

stintel replied:
I noticed, but it's an LTS tag and I did not see those getting used in OpenWrt before. I'll test if include/trusted-firmware-a.mk supports them. Should have done that already.

stintel replied:
Added a change to support A-TF LTS versions. WDYT?

blocktrron commented:
Run-tested on NanoPi R2S

stintel committed:
rkbin: drop TFA_MAKE_FLAGS

The rkbin package doesn't compile anything, so TFA_MAKE_FLAGS is unused.

Suggested-by: Tianling Shen <cnsztl@immortalwrt.org>
Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel committed:
include/trusted-firmware-a.mk: support LTS releases

TF-A LTS releases have an lts- prefix in their version tag. Introduce a
PKG_LTS variable to support these.

As the non-LTS tarbals do not contain the version prefix in the
directory, we need different PKG_BUILD_DIR for each variant:

➜ tar --list --file dl/trusted-firmware-a-v2.12.tar.gz  | head -n1
trusted-firmware-a-2.12/
➜ tar --list --file dl/trusted-firmware-a-lts-v2.12.1.tar.gz  | head -n1
trusted-firmware-a-lts-v2.12.1/

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

1715173329 commented on package/boot/rkbin/Makefile:
```diff
@@ -57,22 +57,19 @@ define Trusted-Firmware-A/rk3568-e25
   TPL:=rk35/rk3588_ddr_lp4_2112MHz_lp5_2400MHz_v1.16.bin
 endef
 
+
```
looks like it's an empty line added by mistake?

1715173329 commented on package/boot/rkbin/Makefile:
```diff
@@ -57,22 +57,19 @@ define Trusted-Firmware-A/rk3568-e25
-define Trusted-Firmware-A/rk3588
+define Trusted-Firmware-A/rk3588-tpl
   BUILD_SUBTARGET:=armv8
-  ATF:=rk35/rk3588_bl31_v1.45.elf
```
Please also add condition check for atf copying below
```
ifneq ($(ATF),)
	$(CP) $(PKG_BUILD_DIR)/bin/$(ATF) $(STAGING_DIR_IMAGE)/
endif
```

1715173329 commented:
Will do a runtime test on Rock 5B later today.

1715173329 commented:
Run-tested on Radxa Rock 5B, looks good so far.

hauke commented on include/trusted-firmware-a.mk:
```diff
@@ -1,12 +1,21 @@
+PKG_LTS ?=
+
+ifneq ($(PKG_LTS),)
+PKG_VERSION_PREFIX:=lts-v
+PKG_BUILD_DIR = $(BUILD_DIR)/$(PKG_NAME)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION_PREFIX)$(PKG_VERSION)
+else
+PKG_VERSION_PREFIX:=v
+PKG_BUILD_DIR = $(BUILD_DIR)/$(PKG_NAME)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION)
+endif
+
 ifndef PKG_SOURCE_PROTO
-PKG_SOURCE = trusted-firmware-a-$(PKG_VERSION).tar.gz
-PKG_SOURCE_URL:=https://codeload.github.com/TrustedFirmware-A/trusted-firmware-a/tar.gz/v$(PKG_VERSION)?
+PKG_SOURCE = trusted-firmware-a-$(_PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/TrustedFirmware-A/trusted-firmware-a/tar.gz/$(PKG_VERSION_PREFIX)$(PKG_VERSION)?
```
Why is this changed to `_PKG_VERSION`?

stintel replied:
Good catch. Must have been a leftover from testing some things. The result is that it downloads the tarball as trusted-firmware-a-.tar.gz. This didn't break because codeload names the file to whatever you pass after the ? Will fix.

hauke commented on include/trusted-firmware-a.mk:
```diff
@@ -1,12 +1,21 @@
+PKG_SOURCE_URL:=https://codeload.github.com/TrustedFirmware-A/trusted-firmware-a/tar.gz/$(PKG_VERSION_PREFIX)$(PKG_VERSION)?
 endif
 
-PKG_BUILD_DIR = $(BUILD_DIR)/$(PKG_NAME)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION)
```
Why not always using:
```
PKG_BUILD_DIR = $(BUILD_DIR)/$(PKG_NAME)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION_PREFIX)$(PKG_VERSION)
```

stintel replied:
This is because the tags are either e.g. lts-v2.12.1 or v2.12. But when downloading the file from Github in a browser, the filenames are:
```
trusted-firmware-a-2.12.tar.gz
trusted-firmware-a-lts-v2.12.1.tar.gz
```
This is reflected in the directory in the tarball:
```
➜ tar --list --file dl/trusted-firmware-a-v2.12.tar.gz  | head -n1
trusted-firmware-a-2.12/
➜ tar --list --file dl/trusted-firmware-a-lts-v2.12.1.tar.gz  | head -n1
trusted-firmware-a-lts-v2.12.1/
```

So we need to specify the v prefix when downloading from codeload, but we need to exclude it in PKG_BUILD_DIR.

stintel replied:
I'll add this to the commit message.

hauke commented on package/boot/uboot-rockchip/Makefile:
```diff
@@ -247,8 +247,10 @@ endef
-  DEPENDS:=+PACKAGE_u-boot-$(1):trusted-firmware-a-rk3588
-  ATF:=rk3588_bl31_v1.45.elf
+  DEPENDS:= \
+	+PACKAGE_u-boot-$(1):trusted-firmware-a-rk3588 \
```
Why does this still depend on `trusted-firmware-a-rk3588`?

stintel replied:
Because that's the name of the package now built from upstream TF-A.

stintel commented:
Thanks for the review folks!
